src/test/lambda/utilities/getMainSourceFile.test.ts

```diff
@@ -342,4 +343,81 @@ describe('getMainSourceFile', async () => {
             assert.fail('Expected an exception, but none was thrown.')
         })
     })
+
+    describe('dotnet', async () => {
+        function createTestTemplate(): CloudFormation.Template {
+            return {
+                Resources: {
+                    HelloWorld: {
+                        Type: 'AWS::Serverless::Function',
+                        Properties: {
+                            Handler: 'HelloWorld::HelloWorld.Function::FunctionHandler',
+                            CodeUri: './artifacts/HelloWorld.zip',
+                            Runtime: 'dotnet'
+                        }
+                    }
+                }
+            }
+        }
+        const expectedHandlerPath = path.join(path.sep, 'dir', 'src', 'HelloWorld', 'Program.cs')
+        it('returns the URI of the main source file for a valid template', async () => {
+            const actual: vscode.Uri = await getMainSourceFileUri({
+                root: vscode.Uri.file('/dir'),
+                getLocalTemplates: (...workspaceUris: vscode.Uri[]) => toAsyncIterable([
+                    vscode.Uri.file('/dir/template.yaml')
+                ]),
+                loadSamTemplate: async uri => createTestTemplate(),
+                fileExists: async p => path.extname(p) === '.cs'
+            })
+
+            assert.strictEqual(actual.fsPath, expectedHandlerPath)
+        })
+
+        it('recognizes all Dotnet runtimes', async () => {
+            async function test(runtime?: string): Promise<void> {
+                const templateUri = vscode.Uri.file(path.join('/dir', 'template.yaml'))
+                const actual: vscode.Uri = await getMainSourceFileUri({
+                    root: vscode.Uri.file('/dir'),
+                    getLocalTemplates: (...workspaceUris: vscode.Uri[]) => toAsyncIterable([templateUri]),
+                    loadSamTemplate: async uri => {
+                        const template = createTestTemplate()
+                        template.Resources!.HelloWorld!.Properties!.Runtime = runtime
+
+                        return template
+                    },
+                    fileExists: async p => path.extname(p) === '.cs'
+                })
+
+                assert.strictEqual(actual.fsPath, expectedHandlerPath)
+            }
+
+            await test('dotnet')
+            await test('dotnetcore1.0')
+            await test('dotnetcore2.0')
+            await test('dotnetcore2.1')
```

Comment:
An issue with a test that internally performs multiple tests is that the whole test stops on first failure. Ideally you have feedback on each test's pass/fail, regardless of other test results. Please adjust to something [closer to parameterized tests](https://mochajs.org/#dynamically-generating-tests) - this is something we should be doing moving forward.

Reply:
The tests for Nodejs, Python and Dotnet follow the same flow. Is your belief that these tests were never OK but that now would be a good time to refactor all of these tests?

Given that `getMainSourceFile` is only used to open the the handler source file after creating a new SAM app and that SAM always creates the same handler file... do you think that perhaps we could consider simplifying all of the is by replacing it with a map `{[runtimename: string]: handlerFilePath: string}`?

Reply:
It was an oversight in reviews until now. Ideally all of the tests would be made parameterized where it makes sense. Ideally each test verifies one concept/input/... . Leave it as-is if you like.